cpp/src/slice2py/PythonUtil.cpp

```diff
@@ -2552,53 +2521,134 @@ namespace
+        try
+        {
+            preprocessor = Preprocessor::create(programName, fileName, preprocessorArgs);
+            FILE* preprocessedHandle = preprocessor->preprocess("-D__SLICE2PY__");
 
-        UnitPtr unit = Unit::createUnit("python", debug);
-        int parseStatus = unit->parse(fileName, cppHandle, false);
+            unit = Unit::createUnit("python", debug);
+            int parseStatus = unit->parse(fileName, preprocessedHandle, false);
 
-        preprocessor->close();
+            preprocessor->close();
 
-        if (parseStatus == EXIT_FAILURE)
-        {
+            if (parseStatus == EXIT_FAILURE)
+            {
+                status = EXIT_FAILURE;
+            }
+            else
+            {
+                if (dependencyGenerator)
+                {
+                    // Collect the dependencies of the unit.
+                    dependencyGenerator->addDependenciesFor(unit);
+                }
+
+                parseAllDocComments(unit, Slice::Python::pyLinkFormatter);
+                validatePythonMetadata(unit);
+
+                unit->visit(&packageVisitor);
+                unit->visit(&importVisitor);
+
+                if (compilationKind == PythonCompilationKind::All || compilationKind == PythonCompilationKind::Module)
+                {
+                    CodeVisitor codeVisitor{
+                        importVisitor.getRuntimeImports(),
+                        importVisitor.getTypingImports(),
+                        importVisitor.getAllImportNames()};
+                    unit->visit(&codeVisitor);
+
+                    const vector<PythonCodeFragment>& newFragments = codeVisitor.codeFragments();
+                    fragments.insert(fragments.end(), newFragments.begin(), newFragments.end());
+                }
+            }
             unit->destroy();
-            throw runtime_error("Failed to parse Slice file: " + fileName);
         }
+        catch (...)
+        {
+            if (preprocessor)
+            {
+                preprocessor->close();
+            }
 
-        parseAllDocComments(unit, Slice::Python::pyLinkFormatter);
-        validatePythonMetadata(unit);
+            if (unit)
+            {
+                unit->destroy();
+            }
+            throw;
+        }
+    }
+
+    if (compilationKind == PythonCompilationKind::All || compilationKind == PythonCompilationKind::Module)
+    {
+        // Write the imports for each code fragment.
+        ImportsMap runtimeImports = importVisitor.getRuntimeImports();
+        ImportsMap typingImports = importVisitor.getTypingImports();
 
-        unit->visit(&packageVisitor);
-        unit->visit(&importVisitor);
+        for (auto& fragment : fragments)
+        {
+            Python::BufferedOutput out;
+            writeHeader(out);
+            writeImports(runtimeImports[fragment.moduleName], typingImports[fragment.moduleName], out);
+            out << sp;
+            out << fragment.code;
```

Comment:
I think would be clear to write the imports, and the header in the code visitor when we start the fragment.

There is no good reason to delay this.